File: app/routes/vendor_credits.py
# ============================================================================
# Vendor Credits — the AP-side counterpart of a customer credit memo
#
# Issue #129 (CimarronSiteServices). A bill posts DR Expense / CR A/P; a
# vendor credit is its mirror, DR A/P / CR Expense. Applying one to a bill
# posts NOTHING — A/P already moved when the credit was issued, and the
# application only decides which bill it settles. That is exactly how a
# credit memo behaves against an invoice.
#
# The reason this is a document and not a journal entry: a manual JE against
# 2000 moves the general ledger without moving the vendor sub-ledger, so A/P
# aging and the vendor's balance stop agreeing with the control account. The
# reporter named that himself, and it is the same split as #119.
# ============================================================================

from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session, joinedload, selectinload
from sqlalchemy.exc import IntegrityError

from app.database import get_db
from app.models.bills import Bill, BillStatus
from app.models.contacts import Vendor
from app.models.items import Item
from app.models.transactions import Transaction
from app.models.vendor_credits import (
    VendorCredit,
    VendorCreditApplication,
    VendorCreditLine,
    VendorCreditStatus,
)
from app.routes._helpers import clamp_pagination
from app.schemas.vendor_credits import (
    VendorCreditApplicationCreate,
    VendorCreditCreate,
    VendorCreditResponse,
)
from app.services.accounting import (
    _q,
    compute_line_totals,
    create_journal_entry,
    get_ap_account_id,
    reversing_lines,
)
from app.services.inventory_service import get_inventory_asset_account_id
from app.services.closing_date import check_closing_date
from app.services.numbering import next_vendor_credit_number
from app.services.purchase_posting import (
    expense_account_for,
    spread,
    unit_cost_with_tax,
)

router = APIRouter(prefix="/api/vendor-credits", tags=["vendor_credits"])


def _with_vendor_name(vc: VendorCredit) -> VendorCreditResponse:
    resp = VendorCreditResponse.model_validate(vc)
    resp.vendor_name = vc.vendor.name if vc.vendor else None
    return resp


@router.get("", response_model=list[VendorCreditResponse])
def list_vendor_credits(
    vendor_id: int = None,
    status: str = None,
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db),
):
    skip, limit = clamp_pagination(skip, limit)
    q = db.query(VendorCredit).options(
        joinedload(VendorCredit.vendor),
        selectinload(VendorCredit.lines),
    )
    if vendor_id:
        q = q.filter(VendorCredit.vendor_id == vendor_id)
    if status:
        q = q.filter(VendorCredit.status == status)
    rows = q.order_by(VendorCredit.date.desc(), VendorCredit.id.desc())
    rows = rows.offset(skip).limit(limit).all()
    return [_with_vendor_name(vc) for vc in rows]


@router.get("/{vc_id}", response_model=VendorCreditResponse)
def get_vendor_credit(vc_id: int, db: Session = Depends(get_db)):
    vc = (
        db.query(VendorCredit)
        .options(joinedload(VendorCredit.vendor), selectinload(VendorCredit.lines))
        .filter(VendorCredit.id == vc_id)
        .first()
    )
    if not vc:
        raise HTTPException(status_code=404, detail="Vendor credit not found")
    return _with_vendor_name(vc)


@router.post("", response_model=VendorCreditResponse, status_code=201)
def create_vendor_credit(data: VendorCreditCreate, db: Session = Depends(get_db)):
    check_closing_date(db, data.date)

    vendor = db.query(Vendor).filter(Vendor.id == data.vendor_id).first()
    if not vendor:
        raise HTTPException(status_code=404, detail="Vendor not found")

    if data.original_bill_id is not None:
        bill = db.query(Bill).filter(Bill.id == data.original_bill_id).first()
        if not bill:
            raise HTTPException(status_code=404, detail="Bill not found")
        if bill.vendor_id != data.vendor_id:
            raise HTTPException(
                status_code=400,
                detail="That bill belongs to a different vendor.",
            )

    subtotal, tax_amount, total = compute_line_totals(data.lines, data.tax_rate)
    if total <= 0:
        raise HTTPException(
            status_code=400,
            detail=(
                "A vendor credit must be for more than zero. Enter the credit "
                "as a positive amount — the document is what makes it a credit."
            ),
        )

    vc = None
    # Same retry as invoices and credit memos: next_vendor_credit_number is
    # MAX+1 with no row lock, so two concurrent creates can compute the same
    # number and one loses at the UNIQUE constraint on flush.
    for _ in range(10):
        vc = VendorCredit(
            credit_number=next_vendor_credit_number(db),
            vendor_id=data.vendor_id,
            date=data.date,
            original_bill_id=data.original_bill_id,
            ref_number=data.ref_number,
            subtotal=subtotal,
            tax_rate=data.tax_rate,
            tax_amount=tax_amount,
            total=total,
            balance_remaining=total,
            notes=data.notes,
            class_id=data.class_id,
            job_id=data.job_id,
            status=VendorCreditStatus.ISSUED,
        )
        db.add(vc)
        try:
            db.flush()
            break
        except IntegrityError as e:
            if "credit_number" not in str(e.orig).lower():
                raise
            db.rollback()
            vc = None
    if vc is None:
        raise HTTPException(
            status_code=503,
            detail="Could not assign a vendor credit number — please retry.",
        )

    journal_lines = []
    returns = []  # [(item, quantity, unit_cost), ...] for inventory lines

    # Round per line so the stored line amount matches compute_line_totals
    # and the credits land on the same cents as the rounded A/P debit.
    amounts = [
        _q(Decimal(str(ln.quantity)) * Decimal(str(ln.rate))) for ln in data.lines
    ]
    # The mirror of a bill: tax on a purchase was part of what the lines
    # cost, so the tax coming back reduces those same lines. It never
    # credits Sales Tax Payable, which is the tax owed on sales.
    tax_shares = spread(tax_amount, amounts)

    for i, line_data in enumerate(data.lines):
        amt = amounts[i]
        item = None
        if line_data.item_id:
            item = db.query(Item).filter(Item.id == line_data.item_id).first()

        if item is not None and item.track_inventory:
            # Returning stock to the vendor takes it off the shelf, so the
            # credit goes against the Inventory asset, not an expense — the
            # exact mirror of how a bill receives it. Refuse rather than
            # silently posting to an expense account, which is what
            # create_bill does for the same reason.
            posting_acct = get_inventory_asset_account_id(db, item)
            if not posting_acct:
                raise HTTPException(
                    status_code=400,
                    detail=(
                        f"Item '{item.name}' is inventory-tracked but has no "
                        "asset_account_id and no account #1300 (Inventory) is "
                        "seeded. Either set the item's inventory asset account "
                        "or add the Inventory account to the chart of accounts."
                    ),
                )
            if line_data.quantity > 0:
                returns.append(
                    (
                        item,
                        Decimal(str(line_data.quantity)),
                        unit_cost_with_tax(
                            amt, tax_shares[i], line_data.quantity, line_data.rate
                        ),
                    )
                )
        elif amt > 0 or line_data.account_id:
            posting_acct = expense_account_for(
                db,
                line_no=i + 1,
                description=line_data.description,
                account_id=line_data.account_id,
                item=item,
                vendor=vendor,
            )
        else:
            posting_acct = (item.expense_account_id if item else None) or (
                vendor.default_expense_account_id
            )

        db.add(
            VendorCreditLine(
                vendor_credit_id=vc.id,
                item_id=line_data.item_id,
                account_id=posting_acct,
                description=line_data.description,
                quantity=line_data.quantity,
                rate=line_data.rate,
                amount=amt,
                job_id=line_data.job_id,
                class_id=line_data.class_id,
                cost_code_id=line_data.cost_code_id,
                line_order=line_data.line_order or i,
            )
        )

        if amt > 0 and posting_acct:
            journal_lines.append(
                {
                    "account_id": posting_acct,
                    "debit": Decimal("0"),
                    "credit": amt + tax_shares[i],
                    "description": line_data.description or "",
                    "job_id": line_data.job_id,
                    "class_id": line_data.class_id,
                    "cost_code_id": line_data.cost_code_id,
                }
            )

    ap_id = get_ap_account_id(db)
    journal_lines.append(
        {
            "account_id": ap_id,
            "debit": total,
            "credit": Decimal("0"),
            "description": f"Vendor Credit {vc.credit_number}",
        }
    )
    txn = create_journal_entry(
        db,
        data.date,
        f"Vendor Credit {vc.credit_number} - {vendor.name}",
        journal_lines,
        source_type="vendor_credit",
        source_id=vc.id,
        reference=vc.ref_number,
        class_id=vc.class_id,
        job_id=vc.job_id,
    )
    vc.transaction_id = txn.id

    if returns:
        from app.models.items import MovementType
        from app.services.inventory_service import _append_movement

        for item, qty, unit_cost in returns:
            # Negative quantity: the goods went back to the vendor. Costed at
            # the credit's own rate, not avg_cost, so a bill and the credit
            # that reverses it net to zero on both quantity and value.
            _append_movement(
                db,
                item,
                MovementType.RETURN_OUT,
                quantity=-qty,
                unit_cost=unit_cost,
                source_type="vendor_credit",
                source_id=vc.id,
                transaction_id=txn.id,
                memo=f"Return to vendor: {vc.credit_number}",
            )

    db.commit()
    db.refresh(vc)
    return _with_vendor_name(vc)


@router.post("/{vc_id}/apply")
def apply_vendor_credit(
    vc_id: int, data: VendorCreditApplicationCreate, db: Session = Depends(get_db)
):
    """Settle part of a bill with a credit the vendor already gave you.

    Posts nothing. A/P moved when the credit was issued; this decides which
    bill it pays down, and both sub-ledger rows move together so the aging
    stays tied to account 2000.
    """
    # Lock both rows for the read-check-write. Two concurrent applies would
    # otherwise both read the same balance_remaining and both pass the check,
    # spending the credit twice. No-op on SQLite; a real lock on Postgres.
    vc = (
        db.query(VendorCredit)
        .filter(VendorCredit.id == vc_id)
        .with_for_update()
        .first()
    )
    if not vc:
        raise HTTPException(status_code=404, detail="Vendor credit not found")
    if vc.status == VendorCreditStatus.VOID:
        raise HTTPException(status_code=400, detail="Vendor credit is voided")

    bill = db.query(Bill).filter(Bill.id == data.bill_id).with_for_update().first()
    if not bill:
        raise HTTPException(status_code=404, detail="Bill not found")
    if bill.status == BillStatus.VOID:
        raise HTTPException(status_code=400, detail="That bill is voided")
    if bill.vendor_id != vc.vendor_id:
        raise HTTPException(
            status_code=400,
            detail=(
                "That bill belongs to a different vendor. A credit can only "
                "be applied to bills from the vendor who issued it."
            ),
        )

    amount = Decimal(str(data.amount))
    if amount <= 0:
        raise HTTPException(status_code=400, detail="Amount must be more than zero")
    if amount > Decimal(str(vc.balance_remaining or 0)):
        raise HTTPException(status_code=400, detail="Amount exceeds credit balance")
    if amount > Decimal(str(bill.balance_due or 0)):
        raise HTTPException(status_code=400, detail="Amount exceeds bill balance")

    db.add(
        VendorCreditApplication(
            vendor_credit_id=vc.id,
            bill_id=data.bill_id,
            amount=amount,
        )
    )

    vc.amount_applied = Decimal(str(vc.amount_applied or 0)) + amount
    vc.balance_remaining = Decimal(str(vc.balance_remaining or 0)) - amount
    if vc.balance_remaining == 0:
        vc.status = VendorCreditStatus.APPLIED

    bill.amount_paid = Decimal(str(bill.amount_paid or 0)) + amount
    bill.balance_due = Decimal(str(bill.balance_due or 0)) - amount
    bill.status = BillStatus.PAID if bill.balance_due == 0 else BillStatus.PARTIAL

    db.commit()
    return {
        "message": f"Applied {amount} to bill {bill.bill_number}",
        "credit_balance_remaining": str(vc.balance_remaining),
        "bill_balance_due": str(bill.balance_due),
    }


@router.post("/{vc_id}/void", response_model=VendorCreditResponse)
def void_vendor_credit(vc_id: int, db: Session = Depends(get_db)):
    """Reverse a vendor credit: put every application back on its bill, post
    the mirror-image entry, put returned stock back on the shelf, and mark it
    void."""
    from app.models.items import MovementType
    from app.services.inventory_service import _append_movement

    vc = (
        db.query(VendorCredit)
        .filter(VendorCredit.id == vc_id)
        .with_for_update()
        .first()
    )
    if not vc:
        raise HTTPException(status_code=404, detail="Vendor credit not found")
    if vc.status == VendorCreditStatus.VOID:
        raise HTTPException(status_code=400, detail="Vendor credit is already void")
    check_closing_date(db, vc.date)

    for app_row in list(vc.applications):
        bill = (
            db.query(Bill).filter(Bill.id == app_row.bill_id).with_for_update().first()
        )
        if bill:
            amt = Decimal(str(app_row.amount))
            bill.amount_paid = Decimal(str(bill.amount_paid or 0)) - amt
            bill.balance_due = Decimal(str(bill.balance_due or 0)) + amt
            bill.status = (
                BillStatus.PARTIAL if bill.amount_paid > 0 else BillStatus.UNPAID
            )
        db.delete(app_row)

    if vc.transaction_id:
        txn = db.query(Transaction).filter(Transaction.id == vc.transaction_id).first()
        if txn is not None:
            create_journal_entry(
                db,
                vc.date,
                f"VOID Vendor Credit {vc.credit_number}",
                reversing_lines(txn.lines),
                source_type="vendor_credit_void",
                source_id=vc.id,
                class_id=vc.class_id,
                job_id=vc.job_id,
            )

    for line in vc.lines:
        if not line.item_id:
            continue
        item = db.query(Item).filter(Item.id == line.item_id).first()
        if item and item.track_inventory and line.quantity > 0:
            _append_movement(
                db,
                item,
                MovementType.VOID,
                quantity=Decimal(str(line.quantity)),
                unit_cost=Decimal(str(line.rate or 0)),
                source_type="vendor_credit_void",
                source_id=vc.id,
                memo=f"VOID Vendor Credit {vc.credit_number}",
            )

    vc.amount_applied = Decimal("0")
    vc.balance_remaining = Decimal("0")
    vc.status = VendorCreditStatus.VOID
    db.commit()
    db.refresh(vc)
    return _with_vendor_name(vc)

File: app/static/js/vendor_credits.js
/**
 * Vendor Credits — the AP-side counterpart of a credit memo (issue #129).
 *
 * A supplier credits you for returned or short-shipped materials. The
 * document debits Accounts Payable and credits the expense (or Inventory,
 * for stock going back). Applying it to a bill posts nothing — it decides
 * which bill the credit settles.
 */
const VendorCreditsPage = {
    async render() {
        const credits = await API.get('/vendor-credits');
        return renderListPage({
            title: 'Vendor Credits',
            headerHtml: `<button class="btn btn-primary" onclick="VendorCreditsPage.showForm()">+ New Vendor Credit</button>`,
            empty: '<p>No vendor credits yet. Enter one when a supplier credits you for a return, a short shipment or an overcharge.</p>',
            columns: ['#', 'Vendor', 'Date', 'Ref', 'Status',
                { label: 'Total', cls: 'amount' }, { label: 'Remaining', cls: 'amount' }, 'Actions'],
            items: credits,
            row: c => `<tr>
                    <td><strong>${escapeHtml(c.credit_number)}</strong></td>
                    <td>${escapeHtml(c.vendor_name || '')}</td>
                    <td>${formatDate(c.date)}</td>
                    <td>${escapeHtml(c.ref_number || '')}</td>
                    <td>${statusBadge(c.status)}</td>
                    <td class="amount">${formatCurrency(c.total)}</td>
                    <td class="amount">${formatCurrency(c.balance_remaining)}</td>
                    <td class="actions">
                        ${c.status === 'issued' ? `<button class="btn btn-sm btn-primary" onclick="VendorCreditsPage.showApply(${c.id})">Apply</button>` : ''}
                        ${c.status !== 'void' ? `<button class="btn btn-sm btn-secondary" onclick="VendorCreditsPage.void(${c.id})">Void</button>` : ''}
                    </td>
                </tr>`,
        });
    },

    async view(id) {
        const c = await API.get(`/vendor-credits/${id}`);
        const lines = (c.lines || []).map(l => `<tr>
                <td>${escapeHtml(l.description || '')}</td>
                <td class="amount">${l.quantity}</td>
                <td class="amount">${formatCurrency(l.rate)}</td>
                <td class="amount">${formatCurrency(l.amount)}</td>
            </tr>`).join('');
        openModal(`Vendor Credit ${escapeHtml(c.credit_number)}`, `
            <p>${escapeHtml(c.vendor_name || '')} &middot; ${formatDate(c.date)} &middot; ${statusBadge(c.status)}</p>
            <div class="table-container"><table>
                <thead><tr><th scope="col">Description</th><th scope="col" class="amount">Qty</th><th scope="col" class="amount">Rate</th><th scope="col" class="amount">Amount</th></tr></thead>
                <tbody>${lines}</tbody>
            </table></div>
            <p style="margin-top:8px;">Total ${formatCurrency(c.total)} &middot; applied ${formatCurrency(c.amount_applied)} &middot; remaining <strong>${formatCurrency(c.balance_remaining)}</strong></p>
            <div class="form-actions"><button class="btn btn-secondary" onclick="closeModal()">Close</button></div>`);
        return '';
    },

    async void(id) {
        if (!confirm('Void this vendor credit? Any applied credit goes back onto the bill, a reversing entry is posted, and returned stock goes back on the shelf.')) return;
        try {
            await API.post(`/vendor-credits/${id}/void`, {});
            toast('Vendor credit voided');
            App.navigate('#/vendor-credits');
        } catch (err) { toast(err.message, 'error'); }
    },

    _items: [],
    _accounts: [],
    _vendors: [],
    lineCount: 0,

    async showForm() {
        const [vendors, items, accounts] = await Promise.all([
            API.get('/vendors?active_only=true'),
            API.get('/items?active_only=true'),
            API.get('/accounts?active_only=true'),
        ]);
        VendorCreditsPage._items = items;
        VendorCreditsPage._vendors = vendors;
        VendorCreditsPage._accounts = PurchaseAccounts.filter(accounts);
        VendorCreditsPage.lineCount = 1;
        const classGroup = await classFormGroupHtml();

        const vendOpts = vendors.map(v => `<option value="${v.id}">${escapeHtml(v.name)}</option>`).join('');

        openModal('New Vendor Credit', `
            <form id="vc-form" onsubmit="VendorCreditsPage.save(event)">
                <p class="form-hint">Enter the credit as a positive amount. It reduces what you owe this vendor, and you choose which bill it settles afterwards.</p>
                <div class="form-grid">
                    <div class="form-group"><label>Vendor *</label>
                        <select name="vendor_id" required><option value="">Select...</option>${vendOpts}</select></div>
                    <div class="form-group"><label>Date *</label>
                        <input name="date" type="date" required value="${todayISO()}"></div>
                    <div class="form-group"><label>Their credit note #</label>
                        <input name="ref_number" placeholder="optional"></div>
                    <div class="form-group"><label>Tax Rate (%)</label>
                        <input name="tax_rate" type="number" step="0.01" min="0" value="0" oninput="VendorCreditsPage.recalc()"></div>
                    ${classGroup}
                </div>
                <h3 style="margin:12px 0 8px;font-size:14px;">Credit Lines</h3>
                <table class="line-items-table">
                    <thead><tr><th scope="col">Item</th><th scope="col">Account</th><th scope="col">Description</th><th scope="col" class="col-qty">Qty</th><th scope="col" class="col-rate">Rate</th><th scope="col" class="col-amount">Amount</th></tr></thead>
                    <tbody id="vc-lines">${VendorCreditsPage.lineHtml(0)}</tbody>
                </table>
                <button type="button" class="btn btn-sm btn-secondary" style="margin-top:8px;" onclick="VendorCreditsPage.addLine()">+ Add Line</button>
                <div class="invoice-totals">
                    <div class="total-row"><span class="label">Subtotal</span><span class="value" id="vc-subtotal">$0.00</span></div>
                    <div class="total-row"><span class="label">Tax</span><span class="value" id="vc-tax">$0.00</span></div>
                    <div class="total-row grand-total"><span class="label">Total</span><span class="value" id="vc-total">$0.00</span></div>
                </div>
                <div class="form-group" style="margin-top:12px;"><label>Notes</label>
                    <textarea name="notes"></textarea></div>
                <div class="form-actions">
                    <button type="button" class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button type="submit" class="btn btn-primary">Create Vendor Credit</button>
                </div>
            </form>`);
    },

    lineHtml(idx) {
        const itemOpts = VendorCreditsPage._items.map(i => `<option value="${i.id}">${escapeHtml(i.name)}</option>`).join('');
        const acctOpts = PurchaseAccounts.options(VendorCreditsPage._accounts);
        return `<tr data-vcline="${idx}">
                <td><select class="line-item" onchange="VendorCreditsPage.itemSelected(this)"><option value="">--</option>${itemOpts}</select></td>
                <td><select class="line-account" aria-label="Account" title="Blank: the item's expense account, else the vendor's default"><option value="">Item / vendor default</option>${acctOpts}</select></td>
                <td><input class="line-desc"></td>
                <td><input class="line-qty" type="number" step="0.01" value="1" oninput="VendorCreditsPage.recalc()"></td>
                <td><input class="line-rate" type="number" step="0.01" value="0" oninput="VendorCreditsPage.recalc()"></td>
                <td class="col-amount line-amount">$0.00</td>
            </tr>`;
    },

    addLine() {
        $('#vc-lines').insertAdjacentHTML('beforeend', VendorCreditsPage.lineHtml(VendorCreditsPage.lineCount++));
    },

    // Picking an item fills what it cost (the rate stayed 0 — W-M1).
    itemSelected(select) {
        const row = select.closest('tr');
        const item = VendorCreditsPage._items.find(i => i.id == select.value);
        if (item) {
            row.querySelector('.line-desc').value = item.description || item.name;
            row.querySelector('.line-rate').value = PurchaseLines.price(item);
        }
        VendorCreditsPage.recalc();
    },

    recalc() {
        let subtotal = 0;
        $$('#vc-lines tr').forEach(row => {
            const amount = PurchaseLines.lineAmount(row);
            subtotal += amount;
            const cell = row.querySelector('.line-amount');
            if (cell) cell.textContent = formatCurrency(amount);
        });
        const pct = parseFloat($('#vc-form [name="tax_rate"]')?.value) || 0;
        const tax = PurchaseLines.cents(subtotal * pct / 100);
        if ($('#vc-subtotal')) $('#vc-subtotal').textContent = formatCurrency(subtotal);
        if ($('#vc-tax')) $('#vc-tax').textContent = formatCurrency(tax);
        if ($('#vc-total')) $('#vc-total').textContent = formatCurrency(subtotal + tax);
        return PurchaseLines.cents(subtotal + tax);
    },

    async save(e) {
        e.preventDefault();
        const form = e.target;
        if (VendorCreditsPage.recalc() <= 0) {
            toast('Enter the amount of the credit: a quantity and rate on at least one line.', 'error');
            return;
        }
        const vendor = VendorCreditsPage._vendors.find(v => v.id == form.vendor_id.value);
        const lines = [];
        let missing = 0;
        $$('#vc-lines tr').forEach((row, i) => {
            const acct = row.querySelector('.line-account')?.value;
            const itemId = row.querySelector('.line-item')?.value;
            const item = itemId ? VendorCreditsPage._items.find(it => it.id == itemId) : null;
            const named = acct || (item && (item.track_inventory || item.expense_account_id)) || (vendor && vendor.default_expense_account_id);
            if (!missing && !named && PurchaseLines.lineAmount(row) > 0) missing = i + 1;
            lines.push({
                item_id: itemId ? parseInt(itemId) : null,
                account_id: acct ? parseInt(acct) : null,
                description: row.querySelector('.line-desc')?.value || '',
                quantity: parseFloat(row.querySelector('.line-qty')?.value) || 1,
                rate: parseFloat(row.querySelector('.line-rate')?.value) || 0,
                line_order: i,
            });
        });
        if (missing) {
            toast(`Choose an account for line ${missing}: its item has none and the vendor has no default expense account.`, 'error');
            return;
        }
        try {
            await API.post('/vendor-credits', {
                vendor_id: parseInt(form.vendor_id.value),
                date: form.date.value,
                ref_number: form.ref_number.value || null,
                tax_rate: (parseFloat(form.tax_rate.value) || 0) / 100,
                notes: form.notes.value || null,
                class_id: classIdFromForm(form),
                lines,
            });
            toast('Vendor credit created');
            closeModal();
            App.navigate('#/vendor-credits');
        } catch (err) { toast(err.message, 'error'); }
    },

    async showApply(vcId) {
        const vc = await API.get(`/vendor-credits/${vcId}`);
        const bills = await API.get(`/bills?vendor_id=${vc.vendor_id}`);
        const open = bills.filter(b => b.status !== 'void' && b.status !== 'paid' && parseFloat(b.balance_due) > 0);

        let rows = open.map(b => `
            <tr>
                <td>${escapeHtml(b.bill_number)}</td>
                <td>${formatDate(b.date)}</td>
                <td class="amount">${formatCurrency(b.balance_due)}</td>
                <td><input type="number" step="0.01" class="vc-apply-amt" data-bill="${b.id}" value="0" style="width:90px;"></td>
            </tr>`).join('');
        if (!rows) rows = '<tr><td colspan="4">No open bills for this vendor. The credit stays on their account until there is one.</td></tr>';

        openModal(`Apply Credit ${escapeHtml(vc.credit_number)}`, `
            <p style="margin-bottom:8px;">Credit remaining: <strong>${formatCurrency(vc.balance_remaining)}</strong></p>
            <div class="table-container"><table>
                <thead><tr><th scope="col">Bill</th><th scope="col">Date</th><th scope="col" class="amount">Balance</th><th scope="col" class="amount">Apply</th></tr></thead>
                <tbody>${rows}</tbody>
            </table></div>
            <div class="form-actions">
                <button class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                <button class="btn btn-primary" onclick="VendorCreditsPage.doApply(${vcId})">Apply Credit</button>
            </div>`);
    },

    async doApply(vcId) {
        const inputs = $$('.vc-apply-amt');
        let applied = 0;
        for (const input of inputs) {
            const amt = parseFloat(input.value) || 0;
            if (amt > 0) {
                try {
                    await API.post(`/vendor-credits/${vcId}/apply`, {
                        bill_id: parseInt(input.dataset.bill), amount: amt,
                    });
                    applied += amt;
                } catch (err) { toast(err.message, 'error'); return; }
            }
        }
        if (!applied) { toast('Enter an amount to apply', 'error'); return; }
        toast('Credit applied');
        closeModal();
        App.navigate('#/vendor-credits');
    },
};
